hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java

```diff
@@ -172,103 +173,132 @@ public void testMultipart() throws Exception {
+    MODIFIED_SINCE_FUTURE_TS_UNMODIFIED_SINCE_FUTURE_TS(
+        futureTimeStr, futureTimeStr,
+        null),
+    MODIFIED_SINCE_FUTURE_TS_UNMODIFIED_SINCE_UNPARSABLE_TS(
+        futureTimeStr, UNPARSABLE_TIME_STR,
+        null);
+    
+    private final String modifiedTimestamp;
+    private final String unmodifiedTimestamp;
+    private final String errorCode;
+
+    CopyIfTimestampTestCase(String modifiedTimestamp,
+        String unmodifiedTimestamp, String errorCode) {
+      this.modifiedTimestamp = modifiedTimestamp;
+      this.unmodifiedTimestamp = unmodifiedTimestamp;
+      this.errorCode = errorCode;
     }
 
-    // False/ifModifiedSince = afterSourceKeyModificationTime
-    // Null
-    try {
-      uploadPartWithCopy(afterSourceKeyModificationTimeStr, null);
-      fail("testMultipartIfModifiedIsFalse");
-    } catch (OS3Exception ex) {
-      assertEquals(ex.getCode(), S3ErrorTable.PRECOND_FAILED.getCode());
+    @Override
+    public String toString() {
+      return this.name() +
+          " Modified:" + this.modifiedTimestamp
+          + " Unmodified:" + this.unmodifiedTimestamp
+          + " ErrorCode:" + this.errorCode;
     }
-    // False/ifModifiedSince = afterSourceKeyModificationTime
-    // Future
-    try {
-      uploadPartWithCopy(afterSourceKeyModificationTimeStr, futureTimeStr);
-      fail("testMultipartIfModifiedIsFalse");
-    } catch (OS3Exception ex) {
-      assertEquals(ex.getCode(), S3ErrorTable.PRECOND_FAILED.getCode());
-    }
-    // False/ifModifiedSince = afterSourceKeyModificationTime
-    // Unparsable
-    try {
-      uploadPartWithCopy(afterSourceKeyModificationTimeStr,
-          UNPARSABLE_TIME_STR);
-      fail("testMultipartIfModifiedIsFalse");
-    } catch (OS3Exception ex) {
-      assertEquals(ex.getCode(), S3ErrorTable.PRECOND_FAILED.getCode());
-    }
-
   }
-  
   @Test
-  public void testMultipartIfModifiedIsTrueOrInvalid() throws Exception {
-    String[] trueOrInvalidTimes = {beforeSourceKeyModificationTimeStr,
-                                   null, UNPARSABLE_TIME_STR, futureTimeStr};
-
-    for (String ts: trueOrInvalidTimes) {
-      // True/Null/Unparsable/Future
-      // True/ifUnmodifiedSince = afterSourceKeyModificationTimeStr
-      uploadPartWithCopy(ts, afterSourceKeyModificationTimeStr);
-  
-      // True/Null/Unparsable/Future
-      // False/ifUnmodifiedSince = beforeSourceKeyModificationTime
+  public void testMultipartTSHeaders() throws Exception {
+    for (CopyIfTimestampTestCase t : CopyIfTimestampTestCase.values()) {
       try {
-        uploadPartWithCopy(ts, beforeSourceKeyModificationTimeStr);
-        fail("testMultipartIfModifiedIsTrueOrInvalid");
+        uploadPartWithCopy(t.modifiedTimestamp, t.unmodifiedTimestamp);
+        if (t.errorCode != null) {
+          fail("Fail test:" + t);
+        }
       } catch (OS3Exception ex) {
-        assertEquals(ex.getCode(), S3ErrorTable.PRECOND_FAILED.getCode());
+        if (t.errorCode == null) {
```

Comment:
Should this check the ERROR_CODE explicitly?

Reply:
yes, thanks @kerneltime!.  fixed now.